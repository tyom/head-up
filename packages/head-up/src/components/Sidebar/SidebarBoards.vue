<template>
  <div class="SidebarBoards">
    <transition-group
      tag="ul"
      name="boardReveal"
      class="board-list"
    >
      <li
        v-for="(board, idx) in boards"
        :key="idx"
        :class="getBoardListItemClass(board, idx)"
        :title="getBoardTitle(board)"
        class="list-item"
      >
        <transition name="actionsReveal">
          <div v-if="editMode && board.editable" class="board-actions">
            <button
              type="button"
              class="remove-button"
              @click.stop="handleRemoveBoard(board.id)"
            >
              <v-icon name="minus-circle"/>
            </button>
          </div>
        </transition>
        <Board
          :cells="board.cells"
          is-thumb
          class="board"
          @click.native="handleActivateBoard(idx)"
        />
        <div v-if="board.title" class="board-title">
          {{ board.title }}
        </div>
      </li>
    </transition-group>
  </div>
</template>

<script>
import Board from '../Board';

export default {
  components: {
    Board,
  },
  computed: {
    activeIdx() {
      return this.getActiveBoardIdx();
    },
    boards() {
      return this.getBoards();
    },
    editMode() {
      return this.isEditing();
    },
  },
  inject: ['isEditing', 'getBoards', 'getActiveBoardIdx'],
  methods: {
    getBoardTitle(board) {
      return board.title + (board.editable ? '' : ' (read-only)');
    },
    getBoardListItemClass(item, idx) {
      return {
        _active: this.activeIdx > -1 ? idx === this.activeIdx : idx === 0,
        ['_read-only']: !item.editable,
      };
    },
    handleRemoveBoard(boardId) {
      this.$emit('remove', boardId);
    },
    handleActivateBoard(boardIdx) {
      this.$emit('activate', boardIdx);
    },
  },
};
</script>

<style scoped>
.board-list {
  margin: 0;
  padding: 0;
  list-style: none;
  display: grid;
  grid-gap: 1em;
}

.list-item {
  position: relative;
}

.board-title {
  font-size: 0.6em;
  margin: 0.5em -0.5em 0;
  text-align: center;
  transition: 0.1s;
}

.board {
  position: relative;
  border: 1px solid #fff2;
  min-width: 4rem;
  width: 8vw;
  height: 10vh;
  transition: 0.2s;

  &:hover {
    border-color: #fff5;
  }

  @nest ._active & {
    border-color: #fffa;
  }

  &:active {
    border-color: #fff;
  }
}

.board:hover + .board-title {
  color: #fff;
}

._read-only .board::after {
  content: '';
  opacity: 0;
  transition: 0.2s;
  position: absolute;
  top: 2px;
  right: 2px;
  bottom: 2px;
  left: 2px;
  pointer-events: none;
  background: repeating-linear-gradient(
    -45deg,
    rgba(#fff, 0),
    rgba(#fff, 0) 0.5rem,
    rgba(#fff, 0.03) 0.5rem,
    rgba(#fff, 0.03) 1rem
  );

  @nest ._edit & {
    opacity: 1;
  }
}

.toggle {
  position: absolute;
  z-index: 2;
  top: 0;
  bottom: 0;
  left: 0;
  width: 1.6rem;
  border: 0;
  color: inherit;
  outline: 0;
  background-color: #111;
  border-right: 1px solid #fff2;

  &:hover {
    color: #fff;
    background-color: #161616;
  }

  &:active {
    background-color: #142a59;
  }
}

.board-actions {
  position: absolute;
  z-index: 1;
  top: -0.5rem;
  right: -0.5rem;
}

.remove-button {
  border: 0;
  padding: 1px;
  background: #fff;
  color: #740000;
  border-radius: 50%;
  display: flex;
  outline: 0;

  & .fa-icon {
    width: 0.8rem;
    height: 0.8rem;
  }

  &:hover {
    color: #bb0000;
  }

  &:active {
    color: #800000;
  }
}

.board-actions {
  position: absolute;
  top: -0.5rem;
  right: -0.5rem;
}

.boardReveal-enter-active,
.boardReveal-leave-active {
  transition: 0.1s;
}
.boardReveal-enter,
.boardReveal-leave-to {
  opacity: 0;
  transform: translateY(-100%);
}

.actionsReveal-enter-active,
.actionsReveal-leave-active {
  transition: 0.1s;
}
.actionsReveal-enter,
.actionsReveal-leave-to {
  transform: scale(0.1);
}
</style>
