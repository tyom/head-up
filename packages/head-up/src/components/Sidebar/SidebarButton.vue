<template>
  <button
    v-bind="$attrs"
    type="button"
    class="button"
    v-on="$listeners"
  >
    <slot/>
  </button>
</template>

<style scoped>
.button {
  width: 100%;
  padding: 0.7em;
  background: none;
  text-align: center;
  line-height: 1;
  color: inherit;
  border: 0;
  outline: 0;

  &:hover {
    color: #fff;
    background-color: rgba(#fff, 0.05);
  }

  &.focus-visible {
    box-shadow: 0 0 0 1px #fffa;
    background-color: rgba(#fff, 0.05);
  }

  &:active {
    color: #fff;
    animation: bounce-1 0.1s forwards;
    background-color: #2b5691;
  }

  &._active {
    background-color: #20416d;
    color: #fff;
    animation: none;

    &:active {
      background-color: #12253e;
    }
  }
}
</style>
