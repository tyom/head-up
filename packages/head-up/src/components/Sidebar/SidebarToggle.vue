<template>
  <div class="SidebarToggle">
    <button class="toggle" @click="$emit('toggle')">
      <v-icon :name="toggleIconName"/>
    </button>
  </div>
</template>

<script>
export default {
  props: {
    toggled: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    toggleIconName() {
      return this.toggled ? 'chevron-left' : 'ellipsis-v';
    },
  },
};
</script>

<style scoped>
.SidebarToggle {
  overflow: hidden;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  z-index: 2;
  box-shadow: 1px 0 1px rgba(#000, 0.5), inset -1px 0 rgba(#fff, 0.1);
}

.toggle {
  width: 1.6rem;
  height: 100%;
  font-size: 1rem;
  border: 0;
  color: inherit;
  outline: 0;
  background-color: transparent;
  transition: background-color 0.1s;

  &:hover {
    color: #fff;
    background-color: rgba(#000, 0.2);
  }

  &:active {
    background-color: #142a59;
    transform: scaleY(1.2);
  }

  &.focus-visible {
    box-shadow: inset 0 0 0 1px rgba(#fff, 0.3);
  }
}
</style>
