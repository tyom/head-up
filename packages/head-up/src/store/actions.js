import { get } from 'lodash';
import Chance from 'chance';

const chance = new Chance();

export default {
  INIT_BOARDS({ dispatch }, { fromProps = [], fromSlots = [] }) {
    dispatch('UPDATE_SERIALIZED_BOARDS', [...fromProps, ...fromSlots]);
  },
  UPDATE_SERIALIZED_BOARDS({ commit }, boards) {
    commit('setSerializedBoards', boards);
  },
  ACTIVATE_BOARD({ commit }, boardId) {
    commit('setActiveBoard', boardId);
  },
  ADD_BOARD({ dispatch, state }) {
    const getWords = words => chance.sentence({ words });
    const title = getWords(2)
      .replace(/\.$/, '')
      .toLowerCase();
    const id = title.replace(' ', '-');

    const newBoard = {
      id,
      title,
      editable: true,
      cells: [
        {
          title: getWords(4),
        },
      ],
    };

    dispatch('UPDATE_SERIALIZED_BOARDS', [newBoard, ...state.serializedBoards]);
    dispatch('ACTIVATE_BOARD', id);

    if (!state.editMode) {
      dispatch('TOGGLE_EDIT_MODE', true);
    }
  },
  REMOVE_BOARD({ dispatch, state }, boardId) {
    const { serializedBoards, activeBoardId } = state;
    const deletionIndex = serializedBoards.findIndex(x => x.id === boardId);
    const currentIndex = serializedBoards.findIndex(
      x => x.id === activeBoardId,
    );

    let nextId;
    const getBoardIdByIndex = idx => get(serializedBoards, `[${idx}].id`);

    if (deletionIndex === currentIndex) {
      const last = deletionIndex === serializedBoards.length - 1;
      nextId = last
        ? getBoardIdByIndex(currentIndex - 1)
        : getBoardIdByIndex(currentIndex + 1);
    } else {
      nextId = getBoardIdByIndex(currentIndex);
    }

    const updatedBoards = serializedBoards.filter(x => x.id !== boardId);
    dispatch('UPDATE_SERIALIZED_BOARDS', updatedBoards);
    dispatch('ACTIVATE_BOARD', nextId);
  },
  UPDATE_BOARD({ dispatch, state }, { boardId, payload }) {
    const updatedBoards = state.serializedBoards.map(
      board => (board.id === boardId ? payload : board),
    );
    dispatch('UPDATE_SERIALIZED_BOARDS', updatedBoards);
    dispatch('ACTIVATE_BOARD', boardId);
  },
  UPDATE_SETTINGS({ commit }, settings) {
    commit('setSettings', settings);
  },
  TOGGLE_SIDEBAR({ commit, state }, value) {
    const isVisible = typeof value === 'undefined' ? !state.showSidebar : value;
    commit('setSidebarState', isVisible);
  },
  TOGGLE_EDIT_MODE({ commit, state }, value) {
    const isEditing = typeof value === 'undefined' ? !state.editMode : value;
    commit('setEditModeState', isEditing);
  },
  RESET_STATE() {
    localStorage.clear();
    window.location.reload();
  },
};
