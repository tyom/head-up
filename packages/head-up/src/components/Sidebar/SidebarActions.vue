<template>
  <div class="SidebarActions">
    <SidebarButton @click="$emit('settings')">
      <v-icon name="cog"/>
    </SidebarButton>
    <SidebarButton @click="$emit('help')">
      <v-icon name="question"/>
    </SidebarButton>
  </div>
</template>

<script>
import SidebarButton from './SidebarButton';

export default {
  components: {
    SidebarButton,
  },
};
</script>

<style scoped>
.SidebarActions {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 0.25em;
}
</style>
