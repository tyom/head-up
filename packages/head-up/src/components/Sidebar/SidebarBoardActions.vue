<template>
  <div class="SidebarBoardActions">
    <SidebarButton
      class="add-button"
      title="Shortcut: a"
      @click="$emit('add-board')"
    >
      <v-icon name="plus"/>
    </SidebarButton>
    <SidebarButton
      :class="{_active: editMode}"
      class="edit-button"
      title="Shortcut: e"
      @click="$emit('toggle-edit')"
    >
      <v-icon name="edit"/>
    </SidebarButton>
  </div>
</template>

<script>
import SidebarButton from './SidebarButton';

export default {
  components: {
    SidebarButton,
  },
  props: {
    editMode: {
      type: Boolean,
      default: false,
    },
  },
};
</script>

<style scoped>
.SidebarBoardActions {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 0.25em;
}
</style>
