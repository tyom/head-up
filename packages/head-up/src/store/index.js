import Vue from 'vue';
import Vuex from 'vuex';
import state from './state';
import mutations from './mutations';
import actions from './actions';
import plugins from './plugins';

Vue.use(Vuex);

export default {
  state,
  mutations,
  actions,
  plugins,
};
