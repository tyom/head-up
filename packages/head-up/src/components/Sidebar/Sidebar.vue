<template>
  <div class="Sidebar">
    <SidebarToggle
      :toggled="visible"
      @toggle="$emit('toggle', !visible)"
    />
    <transition name="slideLeft">
      <div v-show="visible" class="container">
        <SidebarBoardActions
          @add-board="$emit('board:add')"
          @toggle-edit="$emit('board:edit')"
        />
        <SidebarBoards
          @activate="$emit('board:activate', $event)"
          @remove="$emit('board:remove', $event)"
        />
        <SidebarActions
          @help="$emit('modal:help')"
          @settings="$emit('modal:settings')"
        />
      </div>
    </transition>
  </div>
</template>

<script>
import SidebarToggle from './SidebarToggle';
import SidebarBoards from './SidebarBoards';
import SidebarActions from './SidebarActions';
import SidebarBoardActions from './SidebarBoardActions';

export default {
  components: {
    SidebarToggle,
    SidebarBoards,
    SidebarActions,
    SidebarBoardActions,
  },
  props: {
    visible: {
      type: Boolean,
      default: false,
    },
  },
};
</script>

<style scoped>
.Sidebar {
  padding-left: 1.6rem;
  height: 100%;
  position: relative;
}

.container {
  display: grid;
  grid-gap: 1em;
  height: 100%;
  grid-template-rows: auto 1fr auto;
  overflow-y: auto;
  padding: 1em;
  background-color: #0005;
  color: #999;
  box-shadow: inset -1px 0 rgba(#000, 0.5), 1px 0 rgba(#fff, 0.07);
}

.slideLeft-enter-active,
.slideLeft-leave-active {
  transition: 0.15s ease;
}

.slideLeft-enter,
.slideLeft-leave-to {
  margin-left: calc((10vw + 1em) * -1);
  opacity: 0;
}
</style>
