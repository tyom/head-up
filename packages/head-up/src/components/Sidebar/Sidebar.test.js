import { shallowMount } from '@vue/test-utils';
import store from '../../store';
import Sidebar from './Sidebar';
import SidebarToggle from './SidebarToggle';
import SidebarBoards from './SidebarBoards';
import SidebarBoardActions from './SidebarBoardActions';
import SidebarActions from './SidebarActions';

jest.spyOn(store, 'dispatch');

beforeEach(() => {
  jest.clearAllMocks();
});

test('render opened', () => {
  const wrapper = shallowMount(Sidebar, {
    propsData: {
      visible: true,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render closed', () => {
  const wrapper = shallowMount(Sidebar, {
    propsData: {
      visible: false,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

describe('respond to interactions', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Sidebar);
  });

  test('toggle sidebar', () => {
    wrapper.find(SidebarToggle).vm.$emit('toggle');
    expect(store.dispatch).toHaveBeenCalledWith('TOGGLE_SIDEBAR');
  });

  test('add board', () => {
    wrapper.find(SidebarBoardActions).vm.$emit('add-board');
    expect(store.dispatch).toHaveBeenCalledWith('ADD_BOARD');
  });

  test('edit mode', () => {
    wrapper.find(SidebarBoardActions).vm.$emit('toggle-edit');
    expect(store.dispatch).toHaveBeenCalledWith('TOGGLE_EDIT_MODE');
  });

  test('activate board', () => {
    wrapper.find(SidebarBoards).vm.$emit('activate', 'my-board');
    expect(store.dispatch).toHaveBeenCalledWith('ACTIVATE_BOARD', 'my-board');
  });

  test('remove board', () => {
    wrapper.find(SidebarBoards).vm.$emit('remove', 'my-board');
    expect(store.dispatch).toHaveBeenCalledWith('REMOVE_BOARD', 'my-board');
  });

  test('toggle help overlay', () => {
    wrapper.find(SidebarActions).vm.$emit('help');
    expect(wrapper.emitted('toggle:help')[0]).toBeTruthy();
  });

  test('toggle settings overlay', () => {
    wrapper.find(SidebarActions).vm.$emit('settings');
    expect(wrapper.emitted('toggle:settings')[0]).toBeTruthy();
  });
});
