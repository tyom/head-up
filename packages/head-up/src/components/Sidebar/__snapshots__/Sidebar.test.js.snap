// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render closed 1`] = `
<div class="Sidebar">
  <sidebartoggle-stub title="Shortcut: s"></sidebartoggle-stub>
  <div class="container" style="display: none;" name="slideLeft">
    <sidebarboardactions-stub></sidebarboardactions-stub>
    <sidebarboards-stub boards="" activeid=""></sidebarboards-stub>
    <sidebaractions-stub></sidebaractions-stub>
  </div>
</div>
`;

exports[`render opened 1`] = `
<div class="Sidebar">
  <sidebartoggle-stub toggled="true" title="Shortcut: s"></sidebartoggle-stub>
  <div class="container" name="slideLeft">
    <sidebarboardactions-stub></sidebarboardactions-stub>
    <sidebarboards-stub boards="" activeid=""></sidebarboards-stub>
    <sidebaractions-stub></sidebaractions-stub>
  </div>
</div>
`;
